File: code-road/pickFromHereToMigrate/Mathematics/LEETCODE/202. Happy Number.js
/**
 * @param {number} n
 * @return {boolean}
 */

var isHappy = function (n) {
    const lookup = new Map();

    while (true) {

        let sqSum = 0;
        let d1 = 0;

        while (n) {
            d1 = n % 10;
            sqSum += d1 * d1;
            n = Math.floor(n / 10);
        }

        if (sqSum === 1)
            return true;

        if (lookup.has(sqSum))
            return false;

        else {
            lookup.set(sqSum, true);
            n = sqSum;
        }
    }
};

File: code-road/pickFromHereToMigrate/Mathematics/LEETCODE/202. Happy Number.cpp
#include <iostream>
#include <map>

class Solution
{
public:
    std::map<int, bool> lookup;
    bool isHappy(int n)
    {
        int sqSum = 0, d1 = 0;

        while (n)
        {
            d1 = n % 10;
            sqSum += d1 * d1;
            n /= 10;
        }

        if (sqSum == 1)
            return true;

        if (lookup.count(sqSum))
            return false;
        else
        {

            lookup.insert({sqSum, true});

            return isHappy(sqSum);
        }
    }
};

int main()
{
    Solution obj;

    std::cout << std::boolalpha << obj.isHappy(2);

    return 0;
}